forge/test/operators/pytorch/tm/test_reshape.py

```diff
@@ -0,0 +1,309 @@
+# SPDX-FileCopyrightText: © 2024 Tenstorrent AI ULC
+
+# SPDX-License-Identifier: Apache-2.0
+
+# TODO: Add test plan header here
+
+import torch
+import random
+import os
+
+from typing import List, Dict
+from loguru import logger
+
+from forge.verify.config import VerifyConfig
+
+from forge.verify.value_checkers import AllCloseValueChecker
+
+from test.operators.utils import InputSourceFlags, VerifyUtils
+from test.operators.utils import InputSource
+from test.operators.utils import TestVector
+from test.operators.utils import TestPlan
+from test.operators.utils import TestPlanUtils
+from test.operators.utils import FailingReasons
+from test.operators.utils.compat import TestDevice
+from test.operators.utils import TestCollection
+from test.operators.utils import TestCollectionCommon
+from test.operators.utils import ValueRanges
+
+from test.operators.pytorch.eltwise_unary import ModelFromAnotherOp, ModelDirect, ModelConstEvalPass
+
+
+def generate_target_shape(source_shape, max_dims=4, seed=4):
+    """Generates a meaningful target shape based on the source_shape with a limit on the number of dimensions.
+    Ensures that the target shape is not the same as the source shape and that the number of dimensions
+    does not exceed the max_dims parameter."""
+
+    rng = random.Random(seed)
+
+    total_elements = torch.prod(torch.tensor(source_shape)).item()
+    num_attempts = 0  # To prevent infinite loops
+
+    while True:
+        num_dims = rng.randint(1, max_dims)  # Limit the maximum number of dimensions
+
+        # Attempt to divide total_elements into a random number of factors
+        target_shape = []
+        remaining = total_elements
+        for i in range(num_dims - 1):
+            if remaining == 1:
+                target_shape.extend([1] * (num_dims - len(target_shape)))
+                break
+            dim_size = rng.randint(1, remaining)
```

Comment:
This dim_size calculation is brute force. What do you thing about this: Factor remaining into prime factors and then choose some number of factors `k` to take and form a dim_size. Then choose `k` random factors from the list of factors. 

It should look something like this:

```
def random_factor(remaining):
    """Generate a random factor of the given number."""
    factors = prime_factors(remaining)
    len_factors = len(factors)

    if len_factors == 0:
        return 1  # If the number is 1 or prime, the only factor is 1.

    k = random.randint(1, len_factors)
    chosen_factors = random.sample(factors, k)

    # Calculate the product of the chosen factors to form a random factor
    random_factor = 1
    for f in chosen_factors:
        random_factor *= f

    return random_factor
```

To find prime factors, it would be something like:

```
def prime_factors(n):
    """Return the prime factors of a given number."""
    factors = []
    while n % 2 == 0:
        factors.append(2)
        n //= 2
    
    for i in range(3, int(sqrt(n)) + 1, 2):
        while n % i == 0:
            factors.append(i)
            n //= i
    
    if n > 2:
        factors.append(n)
    
    return factors
```